substrate/client/tracing/src/logging/fast_local_time.rs

```diff
@@ -88,24 +100,59 @@ impl FormatTime for FastLocalTime {
 			let mut cache = cache.borrow_mut();
 
 			// Regenerate the timestamp only at most once each second.
-			if cache.last_regenerated_at != unix_time {
-				let ts = chrono::Local::now();
-				let fractional = (ts.nanosecond() % 1_000_000_000) / 1_000_000;
-				cache.last_regenerated_at = unix_time;
-				cache.last_fractional = fractional;
-				cache.buffer.length = 0;
-
-				write!(
-					&mut cache.buffer,
-					"{:04}-{:02}-{:02} {:02}:{:02}:{:02}.{:03}",
-					ts.year(),
-					ts.month(),
-					ts.day(),
-					ts.hour(),
-					ts.minute(),
-					ts.second(),
-					fractional
-				)?;
+			if cache.last_regenerated_at != unix_time || cache.last_utc != self.utc {
+				cache.last_utc = self.utc;
+
+				if self.utc {
+					let ts = chrono::Utc::now();
+					let fractional = (ts.nanosecond() % 1_000_000_000) / 1_000;
+					cache.last_regenerated_at = unix_time;
+					cache.last_fractional = fractional;
+					cache.buffer.length = 0;
+
+					write!(
+						&mut cache.buffer,
+						"{:04}-{:02}-{:02}T{:02}:{:02}:{:02}.{:06}Z",
+						ts.year(),
+						ts.month(),
+						ts.day(),
+						ts.hour(),
+						ts.minute(),
+						ts.second(),
+						fractional
+					)?;
+				} else {
+					let ts = chrono::Local::now();
+					let fractional = (ts.nanosecond() % 1_000_000_000) / 1_000_000;
+					cache.last_regenerated_at = unix_time;
+					cache.last_fractional = fractional;
+					cache.buffer.length = 0;
+
+					write!(
+						&mut cache.buffer,
+						"{:04}-{:02}-{:02} {:02}:{:02}:{:02}.{:03}",
+						ts.year(),
+						ts.month(),
+						ts.day(),
+						ts.hour(),
+						ts.minute(),
+						ts.second(),
+						fractional
+					)?;
+
+					if cache.timezone_offset.length == 0 {
+						write!(&mut cache.timezone_offset, "{}", ts.offset())?;
+					}
+				}
+			} else if self.utc {
+				let fractional = elapsed.subsec_micros();
+
+				// Regenerate the fractional part at most once each millisecond.
+				if cache.last_fractional != fractional {
+					cache.last_fractional = fractional;
+					cache.buffer.length = TIMESTAMP_PARTIAL_UTC_LENGTH + 1;
+					write!(&mut cache.buffer, "{:06}Z", fractional)?;
```

Comment:
6 digits for the fractional part seems somewhat excessive; why not use 3 like in the non-UTC case?

The logger's asynchronous anyway so this much precision is practically useless. It'll just take extra space in the logs for no real benefit.

Reply:
The reason is to make it identical to what `tracing-subscriber` does by default. We use Substrate as a library in a bigger application, so having various logging formats interleaved is awkward and undesirable.

Reply:
Okay, that's reasonable if that's the `tracing-subscriber` default.